### ISSUE TITLE
Visualization Sometimes Blank on First View

### ISSUE THREAD

**Issue by jdunkerley** [enso-org/enso#5620]
### Discord username

jdunkerley

### What type of issue is this?

Permanent – Ocurring repeatably

### Is this issue blocking you from using Enso?

- [X] Yes, I can't use Enso because of this issue.

### Is this a regression?

- [X] Yes, previous version of Enso did not have this issue.

### What issue are you facing?

When trying to visualize a table with 1000 rows, when I perform a function on it the attach visualization is blank.

Closing and reopening it fixes the issue.

### Expected behaviour

Previously, while sometimes slow the viz would appear. Now appears not to.

Ideally, a spinner showing it was loading would be helpful (part of #5088)

### How we can reproduce it?

- Import the weekly book club data
- Get the date of birth column
- Open the viz for that column

### Screenshots or screencasts

![image](https://user-images.githubusercontent.com/4699705/218092434-9f3c4519-ec13-486d-92b2-a6f56b1cfbe3.png)

### Enso Version

Develop as of 10/2

### Browser or standalone distribution

Standalone distribution

### Browser Version or standalone distribution

standalone

### Operating System

Windows

### Operating System Version

Microsoft Windows [Version 10.0.22621.1105] Version 22H21

### Hardware you are using

AMD 6800HS 32GB RTX3050

**Comment by farmaazon** [comment]
> Ideally, a spinner showing it was loading would be helpful.

Spinner is a part of #5088

**Comment by MichaelMauderer** [comment]
For me, this first appears on 9ea9fd56e614d8e4372981cb781ec479140545c9 @hubertp

**Comment by hubertp** [comment]
@MichaelMauderer Unlikely to be the real reason. This commit is a build fix.
@farmaazon is this really a gui or rather language-server issue? (You removed the `triage` label)

**Comment by MichaelMauderer** [comment]
I probably need to investigate a bit first to see what the rot cause is. So far, I just bisected through commits from a GUI perspective. So if this is unlikely to be the candidate, I'll either have a closer look at what is happening or bisect with full engine builds next. (Or do both in parallel).

**Comment by sylwiabr** [comment]
@MichaelMauderer please do, we need to find the cause of the issue ASAP and try to fix it as it is making our product very unstable and undemoable

**Comment by farmaazon** [comment]
> @farmaazon is this really a gui or rather language-server issue? (You removed the triage label)

I don't know, I was just asked directly by Sylwia if our team could investigate it this week, so I removed the label to not cutter "triage" board.

**Comment by MichaelMauderer** [comment]
I'm running into a little issue with bisecting the issue with a full engine build:
* I have one machine where I can build the engine, but I cannot reproduce the issue.
* I have one machine where I can repo the issue but not build the engine (due to some glibc version conflicts that would require me to use a different OS)

It might be best for someone with a working engine build, who can reproduce this issue, to try to bisect this further. @hubertp ?

**Comment by hubertp** [comment]
@MichaelMauderer OK, will try to reproduce.

**Comment by farmaazon** [comment]
@hubertp for time of your investigation, I made you an assignee.

**Comment by farmaazon** [comment]
@hubertp any progress on that?

**Comment by hubertp** [comment]
@farmaazon It's next on my list of tasks but my main task is rather large.
I can reproduce, though.

**Comment by jdunkerley** [comment]
The work I have pushed on improving the stability of the viz has fixed this case.

https://user-images.githubusercontent.com/4699705/221250614-41306ff4-a9b1-4e25-959a-034d8afd5693.mp4

**Comment by hubertp** [comment]
@jdunkerley I will be a bit surprised because I can see no `AttachVisualization` request on the first attempt. 
I think the scenario you describe might be similar but not the same. I will verify once it lands.

**Comment by hubertp** [comment]
@jdunkerley So, as I suspected, this hasn't fixed the case described in this ticket.

I can clearly see that there is something odd happening between IDE and Language Server. It looks as if the latter wasn't receiving any request to attach visualization on the very first attempt (later everything works as expected). Note that in your video, you are already doing some operations. This ticket concerns the situation when on the very first attempt one tries to get the visualization.
Also, the problem is not reproducible when one one has a small number of nodes only.

**Comment by hubertp** [comment]
So I've spent some time investigating when this regression was introduced and what is actually causing it.

https://github.com/enso-org/enso/pull/4120 introduced Table Visualization for Tables and that's when we started visualizing data properly. It also looks like when some form of this regression was introduced - triggering the visualization via a spacebar creates this on/off effect:

https://user-images.githubusercontent.com/292128/222818017-9e787d17-12b0-4584-a790-e64cc4669ce0.mp4

Notice that clicking on an icon worked just fine **always**:

https://user-images.githubusercontent.com/292128/222818305-21b93455-07d0-44fc-ad28-05e167babc77.mp4

There was another change that fixed this [flickering](https://github.com/enso-org/enso/pull/5597). Since then, we have this blank view thing on first view. This only appears to happen on projects with more nodes (like bookclub examples).

Now the reason why this was hard to bisect is because of https://github.com/enso-org/enso/pull/5578 which introduces a breaking change between old IDE and new backend. And that change was merged straight after the #4120. So if you want to reproduce this problem you have to build backend together with IDE before that change.

Another reason why I think this one belongs in IDE territory is because backend receives **no** `attachVisualisation` message from IDE when triggering the visualization. There are a bunch of those that we get during initialization which leads me to suspect that IDE performs some kind of eager caching but we have little control over that. Those requests may also slow down the startup but that's for another discussion.
The only message that backend receives when I trigger a spacebar is `text/applyEdit` which obviously does not perform any visualization computation.
Does this help @farmaazon @MichaelMauderer ?

**Comment by hubertp** [comment]
Another explanation might be that those eager `attachVisualisations` are returning something weird to IDE but a) it probably shouldn't cache such blank views b) I'm guessing clicking on the icon wouldn't work but it clearly does. Either way would probably need some discussion on why and what are we doing with them.

**Comment by farmaazon** [comment]
> #4120 introduced Table Visualization for Tables and that's when we started visualizing data properly. It also looks like when some form of this regression was introduced - triggering the visualization via a spacebar creates this on/off effect:
> 
> Notice that clicking on an icon worked just fine **always**:

This was another issue #5223 fixed in #5624 

> Another reason why I think this one belongs in IDE territory is because backend receives **no** `attachVisualisation` message from IDE when triggering the visualization. 

Ok, it's time to another issue reassignment

> Another explanation might be that those eager attachVisualisations are returning something weird to IDE but a) it probably shouldn't cache such blank views

If attach visualization returns error or times out, the visualization should be hidden. I think I've seen it working recently.

**Comment by farmaazon** [comment]
Did an investigation. Found the reason of those "eager" attachments, made a fix #5841 

Anyway, with or without that fix, I cannot reproduce the issue anymore (on newest develop IDE and 3.4 nightly Engine). The scenario from the description does not work for sure. @jdunkerley do you still reproduce it? If yes, could you provide another scenario here?

**Comment by sylwiabr** [comment]
we had this issues on a call with Wojciech and Ned, will try to do the repro

**Comment by hubertp** [comment]
I can still reproduce on `develop` and on a branch with #5841. I'm still seeing `text/applyEdit` and no `attachVisualization` request. 
I used an example with bookclub1

**Comment by farmaazon** [comment]
> I used an example with bookclub1

What do you mean "used an example"? Do you have a project where you can reproduce it?

**Comment by hubertp** [comment]
> What do you mean "used an example"? Do you have a project where you can reproduce it?

Yeap. I can reproduce it with my solution to bookclub 1. Sorry, if my comment sounded too vague.

https://user-images.githubusercontent.com/292128/223717179-549cc415-a449-48c9-b02d-0c8fe1e9ecd3.mp4

Let me double check again with your fix.

**Comment by hubertp** [comment]
@farmaazon I managed to mix the branches somehow. Your fix works!

**Comment by farmaazon** [comment]
> we had this issues on a call with Wojciech and Ned, will try to do the repro

@sylwiabr do you have this reproduction scenario? Or can I close this issue?

**Comment by farmaazon** [comment]
Closing issue as it seem to not reproduce reliably anymore.
